tests/plugins/test_deepspeed_plugin.py

```diff
@@ -717,14 +717,19 @@ def on_train_batch_start(
     trainer = Trainer(
         default_root_dir=tmpdir,
         progress_bar_refresh_rate=0,
-        max_epochs=5,
+        # TODO: this test fails with max_epochs >1 as there are leftover batches per epoch.
+        # there's divergence in how Lightning handles the last batch of the epoch with how DeepSpeed does it.
+        # we step the optimizers on the last batch but DeepSpeed keeps the accumulation for the next epoch
```

Comment:
This is also a problem in master with an odd `limit_train_batches` and `max_epochs=2`

`$ PL_RUNNING_SPECIAL_TESTS=1 pytest -v --capture=no --disable-warnings tests/plugins/test_deepspeed_plugin.py::test_deepspeed_multigpu_stage_2_accumulated_grad_batches`

Reply:
Any ideas ? I am not sure to grasp the problem.

Reply:
The problem is we get

```python
accumulate_grad_batches = 2

epoch, batch_idx, our_global_step, deepspeed_global_step
0, 0, 0, 0
0, 1, 0, 0 <- both step
0, 2, 1, 1
0, 3, 1, 1 <- both step
0, 4, 2, 2 <- is_last_batch=True, lightning normally steps but since deepspeed is used, it doesn't

1, 5, 2, 2 <- deepspeed steps because it doesnt reset the accumulation at the end of epoch
1, 6, 2, 3 <- lightning steps
1, 7, 3, 3
```

As you can see, Lightning and DeepSpeed go out-of-sync

Reply:
Yes, I know. This is expected :)